src/Store.cs

```diff
@@ -28,6 +28,29 @@ internal void AddFuel(ulong fuel)
             }
         }
 
+        internal object? GetData()
+        {
+            var data = Native.wasmtime_context_get_data(handle);
+            if (data == IntPtr.Zero)
+            {
+                return null;
+            }
+
+            return GCHandle.FromIntPtr(data).Target;
+        }
+
+        internal void SetData(object data)
```

Comment:
If I understand this correctly, when you initially create the `Store` without passing data or passing `null`, but later call `SetData`, then the finalizer would not be called and the `GCHandle` would never be freed, because in `wasmtime_store_new`, `null` would have been passed as finalizer.

(That might be a good candidate for a unit test, e.g. similar to the `ExternRefTests.ItCollectsExternRefs` that checks whether `externref` object instances are freed by the GC after disposing the `Store`.)

I think in order to make this work, we would need to always pass the finalizer to `wasmtime_store_new`, and change the finalizer to something like this (as `IntPtr.Zero` would be an invalid GC handle):
```cs
private static readonly Native.Finalizer Finalizer = (p) =>
{
    if (p != IntPtr.Zero)
    {
        GCHandle.FromIntPtr(p).Free();
    }
};
```

Additionally, `SetData` can then be changed to accept `object?`, and call `GCHandle.Alloc()` only if the new data is not `null`. What do you think?

Reply:
Yes totally agree, good catch. This way calling `SetData(null)` will do what you'd expect, and clear the existing store data. I Added tests to check if GC properly takes care of unused store data on `SetData()` and `Store.Dispose()`